idaes/core/solvers/petsc.py

```diff
@@ -1023,3 +1063,127 @@ def from_json(self, pth):
+    )
+    tdisc = ComponentSet(get_discretization_constraint_data(model, time))
+    flattened_problem = {
+        "timeless_variables": regular_vars,
+        "time_variables": time_vars,
+        "timeless_constraints": regular_cons,
+        "time_constraints": time_cons,
+        "discretization_equations": tdisc,
+    }
+    return flattened_problem
+
+
+def get_initial_condition_problem(
+    model,
+    time,
+    initial_time,
+    representative_time=None,
+    initial_constraints=None,
+    initial_variables=None,
+    detect_initial=True,
+    flattened_problem=None,
+):
+    """
+    Solve a DAE problem step by step using the PETSc DAE solver.  This
+    integrates from one time point to the next.
+
+    Args:
+        model (Block): Pyomo model to solve
+        time (ContinuousSet): Time set
+        initial_time (Element of time): The timepoint to return initial condition problem for
+        representative_time (Element of time): A timepoint at which the DAE system is at its "normal" state
+            after the constraints and variables associated with the initial time point
+            have passed. Typically the next element of time after initial_time. Not needed
+            if flattened_problem is provided.
+        initial_constraints (list): Constraints to solve in the initial
+            condition solve step.  Since the time-indexed constraints are picked
+            up automatically, this generally includes non-time-indexed
+            constraints.
+        initial_variables (list): This is a list of variables to fix after the
+            initial condition solve step.  If these variables were originally
+            unfixed, they will be unfixed at the end of the solve. This usually
+            includes non-time-indexed variables that are calculated along with
+            the initial conditions.
+        detect_initial (bool): If True, add non-time-indexed variables and
+            constraints to initial_variables and initial_constraints.
+        flattened_problem (dict): Dictionary returned by get_flattened_problem.
+            If not provided, get_flattened_problem will be called at representative_time.
+
+    Returns (Pyomo Block):
+        Block containing References to variables and constraints used in initial condition
+        problem, ready to be solved as the initial condition problem.
+    """
+    if initial_variables is None:
+        initial_variables = []
+    # list of constraints to add to the initial condition problem
+    if initial_constraints is None:
+        initial_constraints = []
+
+    if flattened_problem is None:
+        if representative_time is None:
+            raise RuntimeError(
+                "The user must supply either the flattened problem or a representative time."
+            )
+        flattened_problem = _get_flattened_problem(
+            model=model, time=time, representative_time=representative_time
+        )
```

Comment:
Why is either a flattened problem or representative time required? Shouldn't this function just use the same defaults that `petsc_dae_by_time_element` uses?

Reply:
I'm trying to avoid having to flatten the problem multiple times. I haven't profiled it, but I suspect it's rather slow. I debated reimplementing `petsc_dae_by_time_element` as an `Object` so that the flattened problem can be cached. Maybe that will happen eventually.

Reply:
I have no problem with a `flattened_problem` argument, I just think if neither it nor representative time is provided, we should default to `_get_flattened_problem(model=model, time=time, representative_time=time.at(2))`.